src/java/com/palantir/cassandra/check/VolumesIntegrityCheck.java

```diff
@@ -0,0 +1,114 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.cassandra.check;
+
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.UUID;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.palantir.cassandra.utils.FileParser;
+import com.palantir.logsafe.Preconditions;
+import org.apache.cassandra.config.DatabaseDescriptor;
+
+/**
+ * Check if the backing volume has been tampered with. While Cassandra supports multiple data drive, we only
+ * use 1 data drive.
+ */
+public final class VolumesIntegrityCheck
+{
+    public static final String VOLUME_METADATA_NAME = "cassandra-metadata.json";
+
+    private final UUID hostId;
+
+    private final FileParser<VolumeMetadata> dataDriveMetadataFileParser;
+
+    private final FileParser<VolumeMetadata> commitLogMetadataFileParser;
+
+    static class VolumeMetadataType extends TypeReference<VolumeMetadata>
+    {
+    }
+
+    @VisibleForTesting
+    VolumesIntegrityCheck(UUID hostId,
+                          FileParser<VolumeMetadata> dataDriveMetadataFileParser,
+                          FileParser<VolumeMetadata> commitLogMetadataFileParser)
+    {
+        this.hostId = hostId;
+        this.dataDriveMetadataFileParser = dataDriveMetadataFileParser;
+        this.commitLogMetadataFileParser = commitLogMetadataFileParser;
+    }
+
+    public void execute()
+    {
+        try
+        {
+            Optional<VolumeMetadata> maybeDataDriveMetadata = dataDriveMetadataFileParser.read();
+            Optional<VolumeMetadata> maybeCommitLogMetadata = commitLogMetadataFileParser.read();
+            check(maybeDataDriveMetadata, maybeCommitLogMetadata);
+
+            // Write order matters here. We treat write to the data drive as
+            // a "locking" operation, i.e the node is in a steady state.
+            writeIfEmpty(commitLogMetadataFileParser, maybeCommitLogMetadata.isPresent());
+            writeIfEmpty(dataDriveMetadataFileParser, maybeDataDriveMetadata.isPresent());
+        }
+        catch (IOException cause)
+        {
+            throw new RuntimeException("Fail to execute check for volume integrity", cause);
+        }
+    }
+
+    private void check(Optional<VolumeMetadata> maybeDataDriveMetadata, Optional<VolumeMetadata> maybeCommitLogMetadata)
+    {
+        if (maybeCommitLogMetadata.isPresent())
+        {
+            Preconditions.checkState(VolumeMetadata.of(hostId).equals(maybeCommitLogMetadata.get()),
+                                     "Expected SystemKeyspace HostId and POD_NAME to match with commitlog");
```

Comment:
Remind me why we don't also check `!maybeDataDriveMetadata.isPresent() 
 || VolumeMetadata.of(hostId).equals(maybeDataDriveMetadata.get())` here? To prevent data pvcs from switching between two C* pods?

Reply:
If we swap data pvcs, then commit log metadata is present, so `Preconditions.checkState(VolumeMetadata.of(hostId).equals(maybeCommitLogMetadata.get())` throws.